docs: update docs to include budgets (#5339)
aws/karpenter-provider-aws: website/content/en/preview/concepts/disruption.md

## Controls

### Disruption Budgets

> You can rate limit Karpenter's disruption through the `NodePool.Spec.Disruption.Budgets`. If undefined, Karpenter will default to one budget with `nodes: 10%`. Budgets will take into account nodes that are being deleted for any reason, and will only block Karpenter from disrupting nodes voluntarily through expiration, drift, emptiness, and consolidation.

Reviewer: ```suggestion
You can rate limit Karpenter's disruption through the NodePool's `spec.disruption.budgets` field. If undefined, Karpenter will default to one budget with `nodes: 10%`. Budgets will consider nodes that are actively being deleted for any reason, and will only block Karpenter from disrupting nodes voluntarily through expiration, drift, emptiness, and consolidation.
```

> #### budget.Duration + budget.Schedule

Reviewer: Should this just be a note or a warning statement rather than a sub-section? I think our depth here is getting pretty deep and I could just see us having the `#### Schedule` and `#### Duration` sections

Author: Good callout.

Duration and Schedule can only be defined together. When omitted, the budget is always active. When defined, the schedule determines when a budget is active, and the duration determines how long from that point the budget is active. 

##### Schedule

> Schedule is a cronjob schedule. Generally, the cron syntax is five space-delimited values with options below. 

Reviewer: Should we use the word "generally" here? This is always true, right? Or am I missing a different way that crons can be written out?

Author: Nope, you can use the @yearly or @daily https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#:~:text=Other%20than%20the%20standard%20syntax%2C%20some%20macros%20like%20%40monthly%20can%20also%20be%20used%3A

Reviewer: Is it worth just calling out the different ways if there is a minimal set? Rather than saying something like "most of the time" and then not offering the alternatives. If the alternative set is too large, that's fine, just seems like the statement is asking for someone asking, "but what about the non-general case?"

Author: I was thinking the link in the markdown doc has this information, so I didn't want to inflate the doc too much with essentially duplicated information. I added the small list and modified the text here.

Follow the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#writing-a-cronjob-spec) for how to follow the cron syntax. 

```bash
# ┌───────────── minute (0 - 59)
# │ ┌───────────── hour (0 - 23)
# │ │ ┌───────────── day of the month (1 - 31)
# │ │ │ ┌───────────── month (1 - 12)
# │ │ │ │ ┌───────────── day of the week (0 - 6) (Sunday to Saturday;
# │ │ │ │ │                                   7 is also Sunday on some systems)
# │ │ │ │ │                                   OR sun, mon, tue, wed, thu, fri, sat
# │ │ │ │ │
# * * * * *
```

{{% alert title="Note" color="primary" %}}

> Timezones are not supported. Most images default to UTC, but it is best practice to ensure this is the case when considering how to define your budgets. 

Reviewer: This statement is confusing to me. Can we not guarantee that we are _always_ using UTC when we are doing the cron syntax? If this isn't always the case and we aren't using the same time format across images, we're going to get customer issues for disrupting when we shouldn't be

Author: From what I can tell, it isn't guaranteed that all images are using UTC. Anywhere I could find, it all pointed to using the default UTC, but I want to have a note here for anyone specifying timezones within their images. Thoughts?

Reviewer: > I could imagine users who are specifying timezone in their image be very surprised if Karpenter logs in that timezone, but then uses UTC for calculations

I don't think this is surprising to anyone, really. I think a better question is probably how upstream handles this for the cronjob-controller. My expectation was that this time was always reasoned in terms of UTC

Reviewer: Discussed offline. The container filesystem itself should contain information on which timezone that Karpenter is running in. By default, this timezone should be set as UTC but this overridable through a volume mount. It makes sense to me that we would just adhere to the container timezone and stick with UTC for the vast majority of use-cases: https://stackoverflow.com/questions/57607381/how-do-i-change-timezone-in-a-docker-container

{{% /alert %}}

##### Duration

> Duration is a metav1.Duration, which allows compound durations with minutes and hours values such as `10h5m` or `30m` or `160h`. Since cron syntax does not accept denominations smaller than minutes, users can only define minutes or hours. 

Reviewer: To me, this is fairly standard syntax to me. How much do we need to explain how to write this out? For the other fields that use duration, it's assumed so I wonder if we can shorten this

Author: I cut down a little bit, but I was thinking two lines couldn't hurt for users coming straight to this. Thoughts?

#### Nodes

> When calculating if a budget will reject a node from disruption, Karpenter will list the total number of nodes provisioned for a NodePool, and the number of those nodes currently being deleted. 

Reviewer: ```suggestion
When calculating if a budget will block nodes from disruption, Karpenter lists the total number of nodes owned by a NodePool, subtracting out the nodes owned by that NodePool that are currently being deleted. If the number of nodes being deleted by Karpenter or any other processes is greater than the number of allowed disruptions, disruption for this node will not proceed.```

> If the budget is configured with a percentage value, such as `20%`, Karpenter will calculate the number of allowed disruptions as `allowed_disruptions = roundup(total * percentage) - total_deleting`. If otherwise defined as a non-percentage value, Karpenter will simply subtract the number of nodes from the total `(total * percentage) - total_deleting`. For multiple budgets in a NodePool, Karpenter will take the minimum value of each of the budgets.

Reviewer: ```suggestion
If the budget is configured with a percentage value, such as `20%`, Karpenter will calculate the number of allowed disruptions as `allowed_disruptions = roundup(total * percentage) - total_deleting`. If otherwise defined as a non-percentage value, Karpenter will simply subtract the number of nodes from the total `(total * percentage) - total_deleting`. For multiple budgets in a NodePool, Karpenter will take the minimum value (most restrictive) across the budgets.
```

For example, the following NodePool with the following three budgets defines:
- During the first 10 minutes of the day, 0 disruptions are allowed, due to the third budget.

> - When the NodePool has 25 or more nodes, only 5 disruptions are allowed, since 20% of a number larger than 25 would result in 6 nodes, due to rounding up, which is larger than 5. 

Reviewer: Personal preference, I think this works better logically if you order the budgets below in the way that you step through the explanation here.

Also, I think that describing how the percentage field works first as a baseline is easer to reason about. Then, you can step into talking about how to create a ceiling for this percentage value, saying that you can set a budget of 5 to block it going any higher. Finally, you can use your example of blocking disruption during certain times using the schedule/duration syntax.

#### Nodes
When calculating if a budget will block nodes from disruption, Karpenter lists the total number of nodes owned by a NodePool, subtracting out the nodes owned by that NodePool that are currently being deleted. If the number of nodes being deleted by Karpenter or any other processes is greater than the number of allowed disruptions, disruption for this node will not proceed.

If the budget is configured with a percentage value, such as `20%`, Karpenter will calculate the number of allowed disruptions as `allowed_disruptions = roundup(total * percentage) - total_deleting`. If otherwise defined as a non-percentage value, Karpenter will simply subtract the number of nodes from the total `(total * percentage) - total_deleting`. For multiple budgets in a NodePool, Karpenter will take the minimum value (most restrictive) of each of the budgets.

> For example, the following NodePool with three budgets defines:

Reviewer: ```suggestion
For example, the following NodePool with three budgets defines the following requirements:
```

Typically when you have lingo with a colon like this, you want the wording to naturally flow into each line.

> For example, the following NodePool with the following three budgets defines:

Reviewer: ```suggestion
For example, the following NodePool with three budgets defines:
```

redundant

> - When the NodePool has less than 25 nodes nodes, 20% disruptions would be allowed. For instance, if there were 19 nodes owned by the NodePool, 4 disruptions would be allowed, rounding up from `19 * .2 = 3.8`.

Reviewer: ```suggestion
- When the NodePool has less than 25 nodes, 20% disruptions would be allowed. For instance, if there were 19 nodes owned by the NodePool, 4 disruptions would be allowed, rounding up from `19 * .2 = 3.8`.
```

Follow the [Kubernetes documentation](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#writing-a-cronjob-spec) for more information on how to follow the cron syntax. 

```bash
# ┌───────────── minute (0 - 59)
# │ ┌───────────── hour (0 - 23)
# │ │ ┌───────────── day of the month (1 - 31)
# │ │ │ ┌───────────── month (1 - 12)
# │ │ │ │ ┌───────────── day of the week (0 - 6) (Sunday to Saturday;
# │ │ │ │ │                                   7 is also Sunday on some systems)
# │ │ │ │ │                                   OR sun, mon, tue, wed, thu, fri, sat
# │ │ │ │ │
# * * * * *
```

{{% alert title="Note" color="primary" %}}
Timezones are not supported. Most images default to UTC, but it is best practice to ensure this is the case when considering how to define your budgets. 
{{% /alert %}}

#### Duration
Duration allows compound durations with minutes and hours values such as `10h5m` or `30m` or `160h`. Since cron syntax does not accept denominations smaller than minutes, users can only define minutes or hours.

{{% alert title="Note" color="primary" %}}

> Duration and Schedule can only be defined together. When omitted, the budget is always active. When defined, the schedule determines when a budget is active, and the duration determines how long from that point the budget is active. 

Reviewer: ```suggestion
Duration and Schedule must be defined together. When omitted, the budget is always active. When defined, the schedule determines a starting point where the budget will begin being enforced and the duration determines how long from that starting point the budget will be enforced.
```

You don't have to use this exact wording but I think saying something like the fact that the schedule is a starting point is important
